control-plane/agents/src/bin/core/volume/snapshot_operations.rs

```diff
@@ -187,11 +194,20 @@ impl ResourceLifecycleWithLifetime for OperationGuardArc<VolumeSnapshot> {
             *prepare_snapshot.completer.lock().unwrap() = Some(result.clone());
         }
 
-        snapshot
+        match snapshot
             .complete_create(result, registry, OnCreateFail::LeaveAsIs)
-            .await?;
-
-        Ok(snapshot)
+            .await
+        {
+            Ok(_) => Ok(snapshot),
+            Err(error) => {
+                // Prune the snapshots if the creation failed.
+                snapshot
+                    .prune(registry, Some(utils::SNAPSHOT_TRANSACTION_PRUNE_LIMIT))
```

Comment:
Perhaps we should also specify that we want to clean only the current transaction, otherwise this will take whichever transaction?